databroker/broker.py

```diff
@@ -880,3 +902,161 @@ def _munge_time(t, timezone):
+                        filled={k: True for k in external_writers.keys()})
+
+                elif name == 'stop':
+                    for data_key, writer in list(writers.items()):
+                        writer.close()
+                        writers.pop(data_key)
+
+                # The mutated fs_doc is inserted into metadatastore.
+                fs_doc.pop('filled', None)
+                fs_doc.pop('_name', None)
+                db.mds.insert(name, fs_doc)
+
+                # The pristine doc is yielded.
+                yield name, doc
+
+        return wrapped_f
+
+    return wrap
+
+
+def event_map(stream_name, data_keys, provenance):
+    """
+    Map a function onto each event in a stream.
+
+    Parameters
+    ----------
+    stream_name : string
+        e.g., 'primary' or 'baseline'
+    data_keys : dict
+        key(s) in event['data'] to apply function to (e.g., 'image') mapped to
+        a dict with any updates to the data key value
+        (e.g., {'shape': [10, 10]}). In the simple case where the shape,
+        datatype, etc. are unchanged, the dict is just empty:
+        ``{'image': {}}`` and the original metadata passed through.
+    proveance : dict
+        metadata about this operation
+    """
+    def outer(f):
+        def inner(stream):
+            run_start_uid = None
+            descriptor_uid = None
+            for name, doc in stream:
+                if name == 'start':
+                    run_start_uid = str(uuid.uuid4())
+                    new_start_doc = dict(uid=run_start_uid,
+                                         time=time.time(),
+                                         parents=[doc['uid']],
+                                         provenance=provenance)
+                    yield 'start', new_start_doc
+
+                elif name == 'descriptor' and doc.get('name') == stream_name:
+                    if run_start_uid is None:
+                        raise RuntimeError("Received EventDescriptor before "
+                                           "RunStart.")
+                    descriptor_uid = doc_or_uid_to_uid(doc)
+                    new_data_keys = dict(doc['data_keys'])
+                    for k, v in new_data_keys.items():
+                        new_data_keys[k].update(v)
+                    new_descriptor = dict(uid=str(uuid.uuid4()),
+                                          time=time.time(),
+                                          run_start=run_start_uid,
+                                          data_keys=new_data_keys,
+                                          name=stream_name)
+                    yield 'descriptor', new_descriptor
+
+                elif (name == 'event' and
+                      doc_or_uid_to_uid(doc['descriptor']) == descriptor_uid):
+                    if run_start_uid is None:
+                        raise RuntimeError("Received Event before RunStart.")
+                    try:
+                        new_event = dict(doc)
```

Comment:
To expand more on this a little. One of the main issues is this dereferenceing flattens the task graph. Take for example a temperature and image based event stream. At the end of the day you need 1/T and corrected images. These are separate processes and can be handled in their own separate nodes. However, propagating the data encourages users to depend on having the extra data in a single stream. This in turn causes reprocessing issues, if we need to reprocess the data for some reason (faulty image processing) we will also need to reperform the 1/T transformation somewhere in the stack, that where is now a sticky question because depending on where in the chain that transformation occurred either we need to propagate it a lot or a little (even more painful if instead of temperature the second data was another set of images with their own corrections)